jupyterhub/app.py

```diff
@@ -2358,6 +2359,195 @@ async def init_api_tokens(self):
+        config data.
+
+        Args:
+            spec (Dict): The spec of service, defined in the config file.
+            from_config (bool, optional): `True` if the service will be created
+            from the config file, `False` if it is created from REST API.
+            Defaults to `True`.
+            domain (Optional[str]): The domain of Hub session. Defaults to None.
+            host (Optional[str]): The host of Hub session. Defaults to None.
+
+        Returns:
+            Optional[Service]: The created service
+        """
+
+        if domain is None:
+            if self.domain:
+                domain = 'services.' + self.domain
+            else:
+                domain = ''
+
+        if host is None:
+            if self.domain:
+                parsed = urlparse(self.subdomain_host)
+                host = f'{parsed.scheme}://services.{parsed.netloc}'
+            else:
+                host = ''
+
+        if 'name' not in spec:
+            raise ValueError('service spec must have a name: %r' % spec)
+
+        name = spec['name']
+        # get/create orm
+        orm_service = orm.Service.find(self.db, name=name)
+        if orm_service is None:
+            # not found, create a new one
+            orm_service = orm.Service(name=name, from_config=from_config)
+            self.db.add(orm_service)
+            if spec.get('admin', False):
+                self.log.warning(
+                    f"Service {name} sets `admin: True`, which is deprecated in JupyterHub 2.0."
+                    " You can assign now assign roles via `JupyterHub.load_roles` configuration."
+                    " If you specify services in the admin role configuration, "
+                    "the Service admin flag will be ignored."
+                )
+                roles.update_roles(self.db, entity=orm_service, roles=['admin'])
+        else:
+            # Do nothing if the config file tries to modify a API-base service
+            # or vice versa.
+            if orm_service.from_config != from_config:
+                return
+        orm_service.admin = spec.get('admin', False)
+
+        self.db.commit()
+        service = Service(
+            parent=self,
+            app=self,
+            base_url=self.base_url,
+            db=self.db,
+            orm=orm_service,
+            roles=orm_service.roles,
+            domain=domain,
+            host=host,
+            hub=self.hub,
+        )
+
+        traits = service.traits(input=True)
+        for key, value in spec.items():
+            if key not in traits:
+                raise AttributeError("No such service field: %s" % key)
+            setattr(service, key, value)
+            orm_service.update_column(key, value)
+
+        if service.api_token:
+            self.service_tokens[service.api_token] = service.name
+        elif service.managed:
+            # generate new token
+            # TODO: revoke old tokens?
```

Comment:
@minrk this comment is copied from the existing code. Since we're refactoring services is there something we can do to resolve this TODO?

Reply:
Let's do that in its own PR. It's finicky because we want to revoke the previous token, not _all_ tokens associated with the service, which means we have to unambiguously identify when that happens.